olp-cpp-sdk-core/tests/logging/MessageFormatterTest.cpp

```diff
@@ -0,0 +1,501 @@
+
+    element2.type = MessageFormatter::ElementType::Line;
+    EXPECT_FALSE( element1 == element2 );
+    EXPECT_TRUE( element1 != element2 );
+
+    element2 = element1;
+    element2.format = "%u";
+    EXPECT_FALSE( element1 == element2 );
+    EXPECT_TRUE( element1 != element2 );
+
+    element2 = element1;
+    element2.limit = 20;
+    ;
+    EXPECT_FALSE( element1 == element2 );
+    EXPECT_TRUE( element1 != element2 );
+}
+
+TEST( MessageFormatterTest, DefaultLevelMap )
+{
+    const MessageFormatter::LevelNameMap& map = MessageFormatter::defaultLevelNameMap( );
+    EXPECT_EQ( "[TRACE]", map[ static_cast< std::size_t >( Level::Trace ) ] );
+    EXPECT_EQ( "[DEBUG]", map[ static_cast< std::size_t >( Level::Debug ) ] );
+    EXPECT_EQ( "[INFO]", map[ static_cast< std::size_t >( Level::Info ) ] );
+    EXPECT_EQ( "[WARN]", map[ static_cast< std::size_t >( Level::Warning ) ] );
+    EXPECT_EQ( "[ERROR]", map[ static_cast< std::size_t >( Level::Error ) ] );
+    EXPECT_EQ( "[FATAL]", map[ static_cast< std::size_t >( Level::Fatal ) ] );
+}
+
+TEST( MessageFormatterTest, DefaultConstructor )
+{
+    MessageFormatter formatter;
+    EXPECT_TRUE( formatter.getElements( ).empty( ) );
+    EXPECT_EQ( MessageFormatter::defaultLevelNameMap( ), formatter.getLevelNameMap( ) );
+    EXPECT_EQ( MessageFormatter::Timezone::Local, formatter.getTimezone( ) );
+}
+
+TEST( MessageFormatterTest, Constructor )
+{
+    std::vector< MessageFormatter::Element > elements
+        = {MessageFormatter::Element( MessageFormatter::ElementType::String, "LOG: " ),
+           MessageFormatter::Element( MessageFormatter::ElementType::Level, "%s " ),
+           MessageFormatter::Element( MessageFormatter::ElementType::Tag, "%s - " ),
+           MessageFormatter::Element( MessageFormatter::ElementType::File, "%30s:", -30 ),
+           MessageFormatter::Element( MessageFormatter::ElementType::Line, "%5u " ),
+           MessageFormatter::Element( MessageFormatter::ElementType::Time, "[%H:%M] " ),
+           MessageFormatter::Element( MessageFormatter::ElementType::Message )};
+
+    MessageFormatter::LevelNameMap levelNameMap
+        = {{{"Trace"}, {"Debug"}, {"Info"}, {"Warning"}, {"Error"}, {"Fatal"}}};
+
+    MessageFormatter formatter( elements, levelNameMap, MessageFormatter::Timezone::Utc );
+    EXPECT_EQ( elements, formatter.getElements( ) );
+    EXPECT_EQ( levelNameMap, formatter.getLevelNameMap( ) );
+    EXPECT_EQ( MessageFormatter::Timezone::Utc, formatter.getTimezone( ) );
+}
+
+TEST( MessageFormatterTest, CreateDefault )
+{
+    MessageFormatter formatter = MessageFormatter::createDefault( );
+    EXPECT_FALSE( formatter.getElements( ).empty( ) );
+    EXPECT_EQ( MessageFormatter::defaultLevelNameMap( ), formatter.getLevelNameMap( ) );
+    EXPECT_EQ( MessageFormatter::Timezone::Local, formatter.getTimezone( ) );
+}
+
+TEST( MessageFormatterTest, SetElements )
+{
+    MessageFormatter formatter;
+
+    std::vector< MessageFormatter::Element > elements
```

Comment:
Same vector as with `MessageFormatterTest.Constructor` test. Consider to move this vector into common place (anonymous namespace).

Reply:
Moving it away will decrease tests' readability.

Reply:
But keeping it in several places will violate the DRY and you could potentially have misaligned containers.

Reply:
DRY is nice but tests readability is nicer :)
Misaligned containers are also not a problem because every container is different and belongs to an isolated test case.

This is the classical case of how to improve the code that looks identical but has a bit of difference. Moving it into a function will result in overgeneralization. Keeping it as it is will annoy with its duplication.